### ISSUE TITLE
message_get_field: fix the API

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#2869]
`message_get_field()` (imap/message.c) is a function for looking up the value of a given message header.  A message header may have zero (i.e. doesn't exist), one, or multiple values.  When it has multiple values, their order is significant.  Some message headers are further restricted by RFC to only permit zero or one value.

The current implementation:

* is named to kind of sound like it only returns one value
* calls a function `extract_one()` in its implementation which itself sounds like it only returns one value (spoilers: it doesn't)
* accepts a "MESSAGE_MULTIPLE" flag, which sounds like it would make it give back all values when this flag is supplied (and just one value otherwise)
* **always gives back a \r\n-separated `struct buf` of ALL values (in the correct order)**
* ... unless the (new!) flag MESSAGE_LAST is provided (in which case it gives back the last value only, correctly)
* the MESSAGE_MULTIPLE flag is never checked
* accepts a MESSAGE_FIELDNAME flag for requesting the header name (as `"header: value"`) be included in the returned value, but the logic for processing this does not take into account the possibility of the result string containing multiple lines/values, and only deals with the start of the string
* accepts various flags for how value(s) should be decoded/encoded, which appear to work correctly, at least within the limitations observed above and in the contexts they're currently used
* can return the raw message _body_ if you ask for the header "rawbody" (is this why it's called `message_get_field` and not `message_get_header`?)

The only reason our header handling is not totally broken everywhere is because most of our crucial header handling code slurps up the raw message header section (either from message file or cache) and does its own parsing, or uses the message_header_foreach API instead, or naively only looks at the returned value up to the first \r and gets away with it in context.

Most of the problems with this API could be resolved if we changed it to give back the data as a strarray_t of the values (each with optional header name prepended, if that is actually a useful feature), and gave it a name like "message_get_header_values".  Caller would be responsible for deciding whether they wish to use the first value (`strarray_nth(x, 0)`), the last value (`strarray_nth(x, -1)`), all values (by iterating), or abort because multiple-values are contextually invalid, and freeing the strarray_t when they're done with it.

It currently returns zero on success, or an IMAP_* error code (always negative) on error.  It might be useful for it to return a positive number-of-values result on success?

If this API were fixed, we could potentially replace a bunch of bespoke header parsing elsewhere, reducing our exposure to the usual C string parsing issues.

We talked about this in the 9/9 conference call and decided to raise an issue about it (so here it is), but haven't committed to spending any time on it yet.

**Comment by rsto** [comment]
@elliefm agreed that this API is confusing, I tagged this issue as an enhancement. This issue hasn't been updated since 5 years, so we might want to rethink if the Github issue tracker is the right place to keep track of it.
